### ISSUE TITLE
Returns nan when one of the entries in the batch is a digital silence triplet

### ISSUE THREAD

**Issue by iver56** [crlandsc/torch-log-wmse-audio-quality#2]
Here is some code that reproduces the issue:

```
import torch
from torch_log_wmse import LogWMSE

loss_function = LogWMSE(audio_length=1, return_as_loss=True)
torch.manual_seed(0)
raw = torch.randn(2, 1, 44100, dtype=torch.float32)
raw[0] = 0.0
est = torch.randn(2, 1, 44100, dtype=torch.float32)
est[0] = 0.0
gt = torch.randn(2, 1, 44100, dtype=torch.float32)
gt[0] = 0.0
loss0 = loss_function(raw[0:1], est[0:1].unsqueeze(2), gt[0:1].unsqueeze(2))  # -73.6827
print(loss0)
assert not torch.isnan(loss0)
loss1 = loss_function(raw[1:2], est[1:2].unsqueeze(2), gt[1:2].unsqueeze(2))  # 2.7946
print(loss1)
assert not torch.isnan(loss1)
loss_combined = loss_function(raw, est.unsqueeze(2), gt.unsqueeze(2))  # nan
print(loss_combined)
assert not torch.isnan(loss_combined)
```

It looks like this part of the code does not add EPS to the batch entry that is all digital zero:
https://github.com/crlandsc/torch-log-wmse/blob/de47dbc18ff30416636cfbdca146da128ae9497b/torch_log_wmse/metric.py#L99

That leads to the problematic division by zero here:
https://github.com/crlandsc/torch-log-wmse/blob/de47dbc18ff30416636cfbdca146da128ae9497b/torch_log_wmse/metric.py#L103

**Comment by crlandsc** [comment]
Thanks for bringing this up!

Do you mean that if some of the values are 0, but not all of them, it will skip over this block and result in some `inf` entires in `scaling_factor`?

If so, I think it would make sense to just do an item by item compare and replace as such:
```
input_rms[input_rms < ERROR_TOLERANCE_THRESHOLD] = ERROR_TOLERANCE_THRESHOLD
```

Is this what you are saying?

**Comment by iver56** [comment]
> Do you mean that if some of the values are 0, but not all of them, it will skip over this block and result in some inf entires in `scaling_factor`?

Yes.

The original implementation handles the digital silence case specifically and simply returns `return np.log(EPS) * SCALER` early. This is similar to the normal return value, `return np.log((differences**2).mean() + EPS) * SCALER`, assuming `differences` is zero.

https://github.com/nomonosound/log-wmse-audio-quality/blob/3ebe31703a8561b18306c7052a58846931eac306/log_wmse_audio_quality/metric.py#L28

But using an `if` and returning early is not possible when doing batched computations, like we want to do here.

> If so, I think it would make sense to just do an item by item compare and replace as such:
> 
> ```
> input_rms[input_rms < ERROR_TOLERANCE_THRESHOLD] = ERROR_TOLERANCE_THRESHOLD
> ```

That's one viable solution, although in the original implementation of logWMSE, ERROR_TOLERANCE_THRESHOLD was for zeroing out small/tolerable differences between the processed audio and the target audio, not for bumping small input_rms values.

https://github.com/nomonosound/log-wmse-audio-quality/blob/3ebe31703a8561b18306c7052a58846931eac306/log_wmse_audio_quality/metric.py#L36

Using `ERROR_TOLERANCE_THRESHOLD` for two different things does not feel ideal. If we choose this approach, I would prefer to create a separate constant for that purpose.

---

Another way to avoid an "exploding" value is to add a small constant to the divisor, for example

`scaling_factor = 1 / (input_rms + RMS_EPS)`

instead of

`scaling_factor = 1 / input_rms`

If we choose this approach, where we add an `RMS_EPS`, we can remove the if that checks if the input is digital silence.

---

Yet another way to avoid the inf --> nan problem is to check for any inf values in scaling_factor and replace them with zero.

I'm not sure which solution is best yet, but I will give it some thought. Do we want to be "inventive", i.e. create a new/improved version of logWMSE, or do we want a "fix" that matches the calculations in the original implementation?

**Comment by iver56** [comment]
Do you want me to create a PR with a proposed patch?

**Comment by crlandsc** [comment]
Ok, thanks for clearing up your thoughts. That would be really helpful if you made a PR with your proposition!

I think it may be worthwhile leaning more toward the "inventive" side to improve the metric. Because I have already had to adapt/improve the metric regarding the early return of `np.log(EPS) * SCALER` in the original metric code:
```
if np.sum(input_rms) == 0:
    return np.log(EPS) * SCALER
scaling_factor = 1 / input_rms
```

Like you mentioned, this does not work for a loss function, as it skips over calculating the loss entirely if the rms is 0.

My gut instinct would be to simply add a `RMS_EPS` value to `input_rms` as you have shown here, `scaling_factor = 1 / (input_rms + RMS_EPS)`, rather than replacing `infs` with 0s in the `scaling_factor`. My thought process is that if the `input_rms` is 0, then the scaling factor *should be* incredibly large rather than 0. If we add the `RMS_EPS` value to the `scaling_factor` calculation, it maintains the large `scaling_factor` for a small `input_rms` but just caps the max value instead of letting it approach `inf`. Whereas if we replace the `scaling_factor` `infs` with 0s, then very small `input_rms` values will yield large `scaling_factor`, but 0 `input_rms` values will create a stark discontinuity with 0 values for `scaling_factor`.

I am open to thoughts on this!

**Comment by iver56** [comment]
Ok, let's do `RMS_EPS` then. I can create a PR
